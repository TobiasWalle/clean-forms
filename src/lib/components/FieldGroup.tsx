import * as React from 'react';
import { FieldGroupContext, useFieldGroupContext } from '../contexts/field-group-context';
import { createPath } from '../utils';

export interface FieldGroupProps {
  name: string;
  accessor?: string;
  children?: React.ReactNode;
}

function _FieldGroup({ name, accessor, children }: FieldGroupProps) {
  const parentContext = useFieldGroupContext();
  const newContext = {
    path: createPath(parentContext.path, accessor || name),
    namespace: createPath(parentContext.namespace, name)
  };
  return (
    <FieldGroupContext.Provider value={newContext}>
      {children}
    </FieldGroupContext.Provider>
  );
}

export const FieldGroup = React.memo(_FieldGroup);
