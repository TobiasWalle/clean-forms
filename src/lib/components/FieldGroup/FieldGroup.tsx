import * as PropTypes from 'prop-types';
import * as React from 'react';
import { createPath } from '../../utils/createPath';

export const fieldGroupContextTypes = {
  groups: PropTypes.arrayOf(PropTypes.string),
};

export interface FieldGroupContext {
  groups?: string[];
}

export interface FieldGroupProps {
  name: string;
}

export interface FieldGroupState {
}

export class FieldGroup extends React.Component<FieldGroupProps, FieldGroupState> {
  public static childContextTypes = fieldGroupContextTypes;
  public static contextTypes = fieldGroupContextTypes;
  public context: FieldGroupContext;

  public render() {
    return this.props.children;
  }

  public getChildContext(): FieldGroupContext {
    return {
      groups: createPath(this.context.groups, this.props.name)
    };
  }
}
