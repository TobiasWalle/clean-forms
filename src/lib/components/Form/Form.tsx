import * as PropTypes from 'prop-types';
import * as React from 'react';
import { updateDeep } from '../../utils';
import { Errors, ValidationDefinition, Validator } from '../../validation';

export type OnValueChange<Model> = (path: string[], value: any) => void;

export interface FormContext<Model> {
  model: Model;
  onValueChange: OnValueChange<Model>;
}

export const formContextTypes = {
  model: PropTypes.object,
  onValueChange: PropTypes.func,
};

export interface FormMeta<Model> {
  errors: Errors<Model> | undefined;
}

export interface FormProps<Model, FormValidation extends ValidationDefinition<Model> = ValidationDefinition<Model>> {
  model: Model;
  validation?: FormValidation;
  onChange?: (model: Model, meta: FormMeta<Model>) => void;
}

export class Form<Model = any, FormValidation extends ValidationDefinition<Model> = any> extends React.Component<FormProps<Model, FormValidation>, {}> {
  public static childContextTypes = formContextTypes;

  public render() {
    return (
      <form>
        {this.props.children}
      </form>
    );
  }

  public getChildContext(): FormContext<Model> {
    return {
      model: this.props.model,
      onValueChange: this.onValueChange
    };
  }

  private onValueChange: OnValueChange<Model> = (name, value) => {
    this.updateModel(name, value);
  }

  private updateModel: OnValueChange<Model> = (path, value) => {
    const { onChange, model } = this.props;
    const newModel: Model = updateDeep(model, path, value);

    const meta = this.getMetaData(newModel);
    onChange && onChange(newModel, meta);
  }

  private getMetaData(model: Model): FormMeta<Model> {
    return {
      errors: this.validate(model)
    };
  }

  private validate(model: Model): Errors<Model> | undefined {
    const {
      validation = {}
    } = this.props;
    const validator = new Validator<Model>();
    return validator.validate(model, validation);
  }
}
