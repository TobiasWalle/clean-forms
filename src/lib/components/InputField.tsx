import * as React from 'react';
import { createField } from '../hocs';
import { getInputProps } from '../utils/getInputProps';

export const InputField = createField<string, JSX.IntrinsicElements['input']>(
  ({ input: fieldInputProps, custom }) => {
    return <input {...getInputProps(fieldInputProps, custom)} />;
  }
);
