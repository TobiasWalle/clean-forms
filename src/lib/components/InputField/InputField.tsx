import * as React from 'react';
import { createField } from '../../hocs';
import { FieldPropsWithoutRender, FieldRenderFunction, } from '../Field/Field';

export type InputValue = number | string;

export interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {
  label?: string;
}

export const renderInput: FieldRenderFunction<InputValue, InputProps> = ({ input: { name, value, onChange }, custom }) => {
  const label = custom && custom.label;
  return <label>
    <div>{label}</div>
    <input {...custom} name={name} value={value} onChange={event => onChange(event.target.value)}/>
  </label>;
};

export const InputField = createField<InputValue, InputProps>(renderInput);
