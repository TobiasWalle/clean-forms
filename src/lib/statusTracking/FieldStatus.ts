export interface FieldStatusArguments {
  dirty: boolean;
  touched: boolean;
}

export class FieldStatus {
  public readonly dirty: boolean;
  public readonly pristine: boolean;
  public readonly touched: boolean;
  public readonly untouched: boolean;

  constructor(args: FieldStatusArguments) {
    const { dirty, touched } = args;

    this.dirty = dirty;
    this.pristine = !dirty;
    this.touched = touched;
    this.untouched = !touched;
  }
}

export const DEFAULT_FIELD_STATUS: FieldStatus = new FieldStatus({ dirty: false, touched: false });

export function cloneFieldStatus<Children = any>(statusToClone: FieldStatus, overrideArgs: Partial<FieldStatusArguments>): FieldStatus {
  return new FieldStatus({
    ...(statusToClone as any),
    ...overrideArgs
  });
}
