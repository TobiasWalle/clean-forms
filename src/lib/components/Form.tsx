import * as React from 'react';
import {
  FormEvent,
  forwardRef,
  memo,
  MutableRefObject,
  Ref,
  RefAttributes,
  useCallback,
  useEffect,
  useImperativeHandle,
  useLayoutEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { FieldContext, FieldContextValue } from '../contexts/fieldContext';
import {
  FormContext,
  FormContextValue,
  OnFieldBlur,
  OnFieldChange,
  OnFieldMount,
  OnFieldUnmount,
  SetArrayGetKey,
} from '../contexts/formContext';
import { useMemorizedPath } from '../hooks';
import { useAsRef } from '../hooks/useAsRef';
import { useControllableState } from '../hooks/useControllableState';
import { useFormReadApi } from '../hooks/useFormReadApi';
import { useShallowMemo } from '../hooks/useShallowMemo';
import {
  fieldPath as createFieldPath,
  getPathAsString,
  Path,
  path as createPath,
} from '../models';
import { cloneFieldStatus, FieldStatusMapping } from '../statusTracking';
import { FieldStatusUpdater } from '../statusTracking/FieldStatusUpdater';
import { FieldRegister, FieldRegisterChanges } from '../utils';
import { StateUpdater } from '../utils/StateUpdater';
import {
  FieldErrorMapping,
  validateModel,
  ValidationDefinition,
} from '../validation';

import { GetKey } from './FieldArrayItems';

export type OnChange<Model> = (value: Model) => void;
export type OnStatusMappingChange = (mapping: FieldStatusMapping) => void;
export type OnSubmit = (event?: FormEvent<HTMLFormElement>) => void;
export type OnErrorsChange = (errors: FieldErrorMapping) => void;

export interface FormRef {
  /** Trigger a form submit */
  submit: () => void;
}

export type FormProps<Model> = {
  value: Model;
  onChange?: OnChange<Model>;

  status?: FieldStatusMapping;
  onStatusChange?: OnStatusMappingChange;

  onSubmit?: OnSubmit;
  onValidSubmit?: OnSubmit;
  onInValidSubmit?: OnSubmit;
  onErrorsChange?: OnErrorsChange;
  validation?: ValidationDefinition<Model>;
  formProps?: JSX.IntrinsicElements['form'];

  /** Strictly check if the name structure matches the form model. Default is true. */
  strict?: boolean;

  children?: React.ReactNode;
} & RefAttributes<FormRef>;

function useIsMounted(): Readonly<MutableRefObject<boolean>> {
  const mounted = useRef(false);
  useLayoutEffect(() => {
    mounted.current = true;
    return () => {
      mounted.current = false;
    };
  }, []);
  return mounted;
}

function _Form<Model = any>(props: FormProps<Model>, ref: Ref<FormRef>) {
  const {
    value,
    onStatusChange,
    validation,
    formProps,
    onChange,
    onSubmit,
    onErrorsChange,
    onInValidSubmit,
    onValidSubmit,
    strict,
  } = props;
  const [status, setStatus] = useControllableState<FieldStatusMapping>({
    initialStateIfNotControlled: {},
    onChange: onStatusChange,
    controlledState: props.status,
  });

  const setValue = useCallback(
    (newModel: Model) => onChange && onChange(newModel),
    [onChange]
  );

  const valueRef = useAsRef(value);
  const valueUpdaterRef = useRef(
    new StateUpdater<Model>(() => valueRef.current, () => isMountedRef.current)
  );
  useEffect(() => {
    valueUpdaterRef.current.registerOnChange(onChange);
  }, [onChange]);

  const statusRef = useAsRef(status);
  const statusUpdaterRef = useRef(
    new StateUpdater<FieldStatusMapping>(
      () => statusRef.current,
      () => isMountedRef.current
    )
  );
  useEffect(() => {
    statusUpdaterRef.current.registerOnChange(setStatus);
  }, [setStatus]);

  // ArrayGetKeyMapping
  const arrayGetKeyMappingRef = useRef<Map<string, GetKey<any>>>(new Map());

  const setArrayGetKey: SetArrayGetKey<Model> = useCallback((path, getKey) => {
    arrayGetKeyMappingRef.current.set(getPathAsString(path), getKey);
  }, []);

  const removeArrayGetKeyFunction = useCallback((path: Path<Model>) => {
    arrayGetKeyMappingRef.current.delete(getPathAsString(path));
  }, []);

  // FieldRegisterRef
  const handleFieldRegisterChanges = useCallback(
    (changes: FieldRegisterChanges): void => {
      let updatedStatus = status;
      changes.registered.forEach(newPath => {
        updatedStatus = fieldStatusUpdaterRef.current.addIfFieldNotExists(
          updatedStatus,
          newPath
        );
      });

      changes.unregistered.forEach(removedPath => {
        updatedStatus = fieldStatusUpdaterRef.current.removeIfFieldExists(
          updatedStatus,
          removedPath
        );
        removeArrayGetKeyFunction(removedPath);
      });

      statusUpdaterRef.current.update(() => updatedStatus);
    },
    [status, removeArrayGetKeyFunction]
  );

  const [fieldRegister] = useState(new FieldRegister());
  useLayoutEffect(() => {
    fieldRegister.addListener(handleFieldRegisterChanges);
    return () => fieldRegister.removeListener(handleFieldRegisterChanges);
  }, [fieldRegister, handleFieldRegisterChanges]);
  // FieldRegisterRef END

  const fieldStatusUpdaterRef = useRef(new FieldStatusUpdater(fieldRegister));

  const isMountedRef = useIsMounted();

  // API
  const fieldErrorMapping = useMemo<FieldErrorMapping>(
    () =>
      validation
        ? validateModel({ model: value, validationDefinition: validation })
        : {},
    [value, validation]
  );

  useEffect(() => {
    onErrorsChange && onErrorsChange(fieldErrorMapping);
  }, [fieldErrorMapping, onErrorsChange]);

  const {
    valid,
    invalid,
    getFieldStatus,
    getFieldError,
    getFieldValue,
  } = useFormReadApi({
    value,
    status,
    validationDefinition: validation,
    fieldErrorMapping,
    strict,
  });

  // Callbacks
  const handleFieldMount: OnFieldMount<Model> = useCallback(
    path => {
      fieldRegister.register(path);
    },
    [fieldRegister]
  );

  const handleFieldUnmount: OnFieldUnmount<Model> = useCallback(
    path => {
      if (!isMountedRef.current) {
        // Cancel if the form is not mounted anymore
        return;
      }
      fieldRegister.unregister(path);
    },
    [fieldRegister, isMountedRef]
  );

  const getFieldStatusRef = useAsRef(getFieldStatus);
  const handleFieldBlur: OnFieldBlur<Model> = useCallback(
    path => {
      const newStatus = cloneFieldStatus(getFieldStatusRef.current(path), {
        touched: true,
      });
      statusUpdaterRef.current.update(oldStatus => ({
        ...oldStatus,
        [getPathAsString(path)]: newStatus,
      }));
    },
    [getFieldStatusRef]
  );

  const handleFieldChange: OnFieldChange<Model> = useCallback(
    (id, path, fieldValue) => {
      valueUpdaterRef.current.updateDeep(getPathAsString(path), fieldValue);

      const updatedStatus = fieldStatusUpdaterRef.current.markAsDirty(
        statusRef.current,
        id
      );

      statusUpdaterRef.current.update(() => updatedStatus);
    },
    [statusRef]
  );
  // Callbacks End

  const markAllAsTouched = useCallback(() => {
    const updatedStatus = fieldStatusUpdaterRef.current.markAllAsTouched(
      statusRef.current
    );
    statusUpdaterRef.current.update(() => updatedStatus);
  }, [statusRef]);

  const submit = useCallback(
    (event?: React.FormEvent<HTMLFormElement>) => {
      markAllAsTouched();
      onSubmit && onSubmit(event);
      if (valid) {
        onValidSubmit && onValidSubmit(event);
      } else {
        onInValidSubmit && onInValidSubmit(event);
      }
    },
    [markAllAsTouched, onSubmit, valid, onValidSubmit, onInValidSubmit]
  );

  const handleSubmit = useCallback(
    (event: React.FormEvent<HTMLFormElement>) => {
      event.stopPropagation();
      event.preventDefault();

      formProps && formProps.onSubmit && formProps.onSubmit(event);

      submit(event);
    },
    [submit, formProps]
  );

  const formContext: FormContextValue<Model> = useShallowMemo({
    valid,
    invalid,
    getFieldStatus,
    getFieldError,
    getFieldValue,
    onFieldBlur: handleFieldBlur,
    onFieldChange: handleFieldChange,
    onFieldMount: handleFieldMount,
    onFieldUnmount: handleFieldUnmount,
    setArrayGetKey,
  });

  const error = getFieldError();
  const markAsTouched = useCallback(() => handleFieldBlur(createPath()), [
    handleFieldBlur,
  ]);

  const rootFieldContext: FieldContextValue<Model> = useShallowMemo({
    fieldPath: useMemorizedPath(createFieldPath<Model>()),
    modelPath: useMemorizedPath(createPath<Model>()),
    name: '',
    value,
    setValue,
    markAsTouched,
    error,
    valid: !error,
    invalid: !!error,
    ...getFieldStatus(),
  });

  useImperativeHandle(ref, () => ({ submit }), [submit]);

  return (
    <FormContext.Provider value={formContext}>
      <FieldContext.Provider value={rootFieldContext}>
        <form {...formProps} onSubmit={handleSubmit}>
          {props.children}
        </form>
      </FieldContext.Provider>
    </FormContext.Provider>
  );
}

export type Form = typeof _Form;
export const Form: Form = memo(forwardRef(_Form)) as any;
