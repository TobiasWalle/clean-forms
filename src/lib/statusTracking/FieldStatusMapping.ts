import { FieldStatus } from './FieldStatus';

export interface FieldStatusMapping {
  [path: string]: FieldStatus;
}
