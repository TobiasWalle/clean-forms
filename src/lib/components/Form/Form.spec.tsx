import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { Field, FieldArray, FieldArrayItems } from '../';
import { InputField } from '../../../testUtils/InputField';
import { wait } from '../../../testUtils/wait';
import { FieldRegister } from '../../utils/FieldRegister';
import { DEFAULT_FIELD_STATUS, FieldStatus } from '../../utils/statusTracking/FieldStatus';
import { FieldStatusMapping } from '../../utils/statusTracking/FieldStatusMapping';
import {
  ArrayValidation, FieldErrorMapping,
  ValidationDefinition
} from '../../utils/validation';
import { FieldGroup } from '../FieldGroup/FieldGroup';
import { Form, FormContext } from './Form';

describe('Form', () => {
  it('should render', () => {
    expect(shallow(<Form state={{ model: {} }}/>)).toMatchSnapshot();
  });

  it('should render model in inputs', () => {
    const model = {
      a: 'hello',
      b: 124
    };
    const element = mount(
      <Form state={{ model }}>
        <InputField name={'a'}/>
        <InputField name={'b'} type="number"/>
      </Form>
    );
    const inputs = element.find('input');
    const firstInputProps = inputs.first().props();
    const secondInputProps = inputs.at(1).props();

    expect(firstInputProps.name).toBe('a');
    expect(firstInputProps.value).toBe('hello');
    expect(secondInputProps.name).toBe('b');
    expect(secondInputProps.value).toBe(124);
    expect(secondInputProps.type).toBe('number');
    expect(element.html()).toMatchSnapshot();
  });

  it('should emit model changes', () => {
    const onChange = jest.fn();
    const model = {
      a: 'hello',
      b: 124
    };
    const status: FieldStatusMapping = initFieldStatusMapping('a');
    const element = mount(
      <Form state={{ model, status }} onChange={onChange}>
        <InputField name={'a'}/>
        <InputField name={'b'} type="number"/>
      </Form>
    );
    const expectedNewValue = 'new';
    const inputs = element.find('input');
    const firstInput = inputs.first();

    firstInput.simulate('change', { target: { value: expectedNewValue } });

    expect(onChange).toHaveBeenCalledWith({
      model: {
        ...model,
        a: expectedNewValue
      },
      status: expect.anything()
    });
  });

  it('should emit model changes in groups', () => {
    const onChange = jest.fn();
    const model = {
      a: 'hello',
      b: 124,
      c: {
        c1: 'old'
      }
    };
    const status: FieldStatusMapping = initFieldStatusMapping('a', 'b', 'c.c1');
    const element = mount(
      <Form state={{ model, status }} onChange={onChange}>
        <InputField name={'a'}/>
        <InputField name={'b'} type="number"/>
        <FieldGroup name={'c'}>
          <InputField name={'c1'}/>
        </FieldGroup>
      </Form>
    );
    const expectedNewValue = 'new';
    const inputs = element.find('input');
    const thirdInput = inputs.at(2);

    thirdInput.simulate('change', { target: { value: expectedNewValue } });

    expect(onChange).toHaveBeenCalledWith({
      model: {
        ...model,
        c: { c1: expectedNewValue }
      },
      status: expect.anything()
    });
  });

  it('should initialize status', () => {
    const onChange = jest.fn();
    const model = { a: 'hello', b: 124 };
    const expectFieldStatus = createFieldStatusExpectFunction(onChange);

    mount(
      <Form state={{ model }} onChange={onChange}>
        <InputField name={'a'}/>
        <InputField name={'b'} type="number"/>
      </Form>
    );
    expectFieldStatus({
      a: new FieldStatus({ dirty: false, touched: false }),
      b: new FieldStatus({ dirty: false, touched: false })
    });
  });

  it('should update dirty/pristine status', async () => {
    const onChange = jest.fn();
    const model = { a: 'hello', b: 124 };
    const status: FieldStatusMapping = initFieldStatusMapping('a', 'b');
    const expectFieldStatus = createFieldStatusExpectFunction(onChange);
    const element = mount(
      <Form state={{ model, status }} onChange={onChange}>
        <InputField name={'a'}/>
        <InputField name={'b'} type="number"/>
      </Form>
    );
    const inputs = element.find('input');
    const firstInput = inputs.first();

    firstInput.simulate('change', { target: { value: '0123456789' } });

    expectFieldStatus({
      a: new FieldStatus({ dirty: true, touched: false }),
      b: DEFAULT_FIELD_STATUS
    });
  });

  it('should update touched/untouched status', async () => {
    const onChange = jest.fn();
    const model = { a: 'hello', b: 124 };
    const status: FieldStatusMapping = initFieldStatusMapping('a', 'b');
    const expectFieldStatus = createFieldStatusExpectFunction(onChange);
    const element = mount(
      <Form state={{ model, status }} onChange={onChange}>
        <InputField name={'a'}/>
        <InputField name={'b'} type="number"/>
      </Form>
    );
    const inputs = element.find('input');
    const firstInput = inputs.first();

    firstInput.simulate('blur', {});

    expectFieldStatus({
      a: new FieldStatus({ dirty: false, touched: true }),
      b: DEFAULT_FIELD_STATUS
    });
  });

  it('should update validation result', async () => {
    const model = { a: 'hello', b: 124 };
    const status = initFieldStatusMapping('a');
    const error = 'The value has to have a greater length than 10';
    const validators: ValidationDefinition<typeof model> = {
      a: ({ value }) => value.length > 10 ? null : error
    };
    const renderForm = jest.fn(() => (
      <>
      <InputField name={'a'}/>
      <InputField name={'b'} type="number"/>
      </>
    ));
    const expectValidationResult = createValidationResultExpectFunction(renderForm);
    mount(<Form state={{ model, status }} validation={validators} render={renderForm}/>);

    expectValidationResult({
      a: error
    });

  });

  it('should update validation result for array', async () => {
    const model = { array: [] };
    const status = initFieldStatusMapping('array', 'array.hello');
    const error = 'The array needs at least 1 item';
    const validators: ValidationDefinition<typeof model> = {
      array: new ArrayValidation(null, ({ value }) => value.length >= 1 ? null : error)
    };
    const renderForm = jest.fn(() => (
      <FieldArray name={'array'} render={() => (
        <FieldArrayItems getKey={item => item} render={() =>
          <InputField name={null}/>
        }/>
      )}/>
    ));
    const expectValidationResult = createValidationResultExpectFunction(renderForm);
    mount(
      <Form state={{ model, status }} validation={validators} render={renderForm}/>
    );
    expectValidationResult({
      array: error
    });
  });

  it('should remove status if field unmounts', async () => {
    const onChange = jest.fn();
    const model = { a: 'test' };
    const expectFieldStatus = createFieldStatusExpectFunction(onChange);
    const element = mount(
      <Form state={{ model, status: { a: DEFAULT_FIELD_STATUS } }} onChange={onChange}>
        <InputField name={'a'}/>
      </Form>
    );
    const context: FormContext<any> = element.find(Field).instance().context;

    context.onFieldUnmount('a');

    await wait(10);

    expectFieldStatus({});
  });

  it('should remove status if array item is removed', async () => {
    const onChange = jest.fn();
    const model = { array: ['a', 'b'] };
    const status: FieldStatusMapping = initFieldStatusMapping('array', 'array.a', 'array.b');
    const expectFieldStatus = createFieldStatusExpectFunction(onChange);
    const element = mount(
      <Form state={{ model, status }} onChange={onChange}>
        <FieldArray name={'array'} render={() => (
          <FieldArrayItems getKey={item => item} render={() => (
            <InputField name={null}/>
          )}/>
        )}/>
      </Form>
    );
    const context: FormContext<any> = element.find(Field).first().instance().context;

    context.onFieldUnmount('array.a');

    await waitForDebounce();

    expectFieldStatus({
      array: DEFAULT_FIELD_STATUS,
      'array.b': DEFAULT_FIELD_STATUS
    });
  });

  it('should trigger onSubmit', () => {
    const model = { a: 123 };
    const onSubmit = jest.fn();
    const element = shallow(
      <Form state={{model}} onSubmit={onSubmit}/>
    );

    const form = element.find('form');
    form.props().onSubmit!(new Event('test') as any);

    expect(onSubmit).toHaveBeenCalledWith({
      state: { model },
      validationDefinition: expect.anything(),
      fieldErrorMapping: expect.anything(),
      valid: true,
      inValid: false
    });
  });
});

function createFieldStatusExpectFunction(onChange: jest.Mock<any>) {
  return (fieldStatusMapping: FieldStatusMapping) => {
    const lastCall = onChange.mock.calls[onChange.mock.calls.length - 1];
    expect(lastCall[0].status).toEqual(fieldStatusMapping);
  };
}

function createValidationResultExpectFunction(renderForm: jest.Mock<any>) {
  return (fieldErrorMapping: FieldErrorMapping) => {
    const lastCall = renderForm.mock.calls[renderForm.mock.calls.length - 1];
    expect(lastCall[0].fieldErrorMapping).toEqual(fieldErrorMapping);
  };
}

function waitForDebounce(): Promise<void> {
  return wait(FieldRegister.DEBOUNCE_IN_MS);
}

function initFieldStatusMapping(...keys: string[]): FieldStatusMapping {
  return keys.reduce((result, key) => {
    result[key] = DEFAULT_FIELD_STATUS;
    return result;
  }, {} as FieldStatusMapping);
}
