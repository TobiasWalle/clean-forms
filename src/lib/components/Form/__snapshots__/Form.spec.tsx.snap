// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render 1`] = `
<form
  onSubmit={[Function]}
/>
`;

exports[`Form should render model in inputs 1`] = `"<form><input name=\\"a\\" value=\\"hello\\"><input type=\\"number\\" name=\\"b\\" value=\\"124\\"></form>"`;
